### ISSUE TITLE
-d:useMalloc for the default GC

### ISSUE THREAD

**Issue by mratsim** [nim-lang/Nim#15394]
At Status, we have been fighting against resource leaks for months (sockets, file descriptors, futures, timers and most importantly memory)
- https://github.com/status-im/nim-libp2p/pull/174
- https://github.com/status-im/nim-libp2p/pull/134
- https://github.com/status-im/nim-libp2p/pull/318
- https://github.com/status-im/nim-libp2p/pull/371
- https://github.com/status-im/nim-eth/commit/6ec942a195d3e3fcdff5e5fa39a458f06f4028aa
- https://github.com/status-im/nim-eth/pull/39
- https://github.com/status-im/nim-chronos/pull/102
- https://github.com/status-im/nim-chronos/pull/130
- https://github.com/status-im/nim-chronos/pull/65
- https://github.com/status-im/nim-chronos/issues/53
- https://github.com/status-im/nim-beacon-chain/pull/1673
- https://github.com/status-im/nim-beacon-chain/pull/1011
- https://github.com/status-im/nim-beacon-chain/issues/850
- https://github.com/status-im/nim-beacon-chain/issues/1629
- https://github.com/status-im/nim-beacon-chain/issues/1123
- https://github.com/status-im/nim-beacon-chain/issues/1121
- https://github.com/status-im/nim-beacon-chain/pull/1474#issuecomment-672043600
- https://github.com/status-im/nim-beacon-chain/issues/1010
- https://github.com/status-im/nim-beacon-chain/issues/447
- https://github.com/status-im/nim-beacon-chain/issues/784
- https://github.com/status-im/nim-beacon-chain/issues/779
- https://github.com/status-im/nim-libp2p/pull/128

We built several tools to help us track the issues
- https://github.com/status-im/nim-libp2p/pull/136
- https://github.com/status-im/nim-libp2p/issues/207
- https://github.com/status-im/nim-chronos/pull/33
- https://github.com/status-im/nim-chronos/issues/31
- https://github.com/status-im/nim-chronos/pull/85
- https://github.com/status-im/nim-chronos/issues/32
- https://github.com/status-im/nim-beacon-chain/issues/869
- https://github.com/status-im/nim-chronos/pull/106

However we are still leaking. As a workaround we are currently advising people to restart every 6 hours or so but for production we need to remove all possible source of leaks so that user can run that application for months without restart.

The computational part of our application is relatively easy to debug for leaks but the async/IO/networking part as been leaking resources quite often via closure iterators/futures due to unattended cancellation/expiration.

This is incredibly hard to debug.

We would like to have ``-d:useMalloc`` available for the default GC, backported to the 1.2.x branch so that we can use conventional C tools like Valgrind to detect those memory leaks and also run memory leak detection in dedicated CI.

Furthermore, we are currently investigating an issue with misreported memory accounting by Nim GC, which makes it even harder to debug with Nim standard tools (memory fragmentation?).

**Comment by Araq** [comment]
That seems to be more work than moving to 1.4's ``--gc:orc``.

**Comment by pmetras** [comment]
Even with 1.4 and --gc:orc, there are situations where resources can leak: https://github.com/nim-lang/Nim/issues/13289.
From what I understand, `--gc:orc` only tries to prevent memory leaks but does not help to deal with other types of leaks.

**Comment by ghost** [comment]
@pmetras wdym? with ORC no memory leaks should happen, otherwise it's a bug in ARC/ORC itself or in code which used manual memory allocation or with custom destructors (if they're not written correctly)

**Comment by pmetras** [comment]
@Yardanico , I agree that ORC solves memory leaks. I just want to say that there are other types of leaks, like listed by mratsim (sockets, file descriptors, futures, timers, BD cursors, OS locks, etc.) and that ORC does not help with these types of leaks.

**Comment by ghost** [comment]
@pmetras but it can and that's one of the reasons why ARC/ORC was created in the first place. You can create a custom destructor for any types including sockets, file descriptors, futures, locks, etc

**Comment by ghost** [comment]
@pmetras just for a simple example (compile with arc/orc for reliable results, or add GC_fullCollect() after main with refc)
```nim
type
  MyFile = object
    f: File

proc `=destroy`(myf: var MyFile) = 
  echo "destroying"
  myf.f.close() # close already does the isNil check for us

proc main = 
  let myf = MyFile(f: open("a.nim"))
  echo myf.f.readAll()

main()
```

I use a wrapper object MyFile here because you can't create destructors for pointers (and File is a "ptr CFile"). So as you can see it's not even hard

**Comment by pmetras** [comment]
@Yardanico, OK I understand how it can work: I'll have to encapsulate resource handles into custom types with a finalizer. I had an example working too but did not post it. Perhaps the `File` type in stdlib will be changed to become autoclosing in 1.6?
Thanks for your example.

**Comment by arnetheduck** [comment]
> That seems to be more work than moving to 1.4's --gc:orc.

The litmus test here is that not even the compiler works with orc/arc by default - Nimbus is quite a bit a larger project that is expected to run over long periods of time without restard, and that that has been tuned and optimized to the current GC - introducing a new GC, a new memory layout for core types like string and seq etc is a major risk given that it will bring new bugs, require new workarounds and require reoptimizing the memory profile from scratch - it's not something we can do frivolously. We're pretty much aware of how orc / arc works but it's not what we need in this particular case.
